src/server/routes/externalService.js

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2022       Ansh Goyal
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+
+import * as auth from '../helpers/auth';
+import * as cbHelper from '../helpers/critiquebrainz';
+import * as propHelpers from '../../client/helpers/props';
+import {AuthenticationFailedError, BadRequestError} from '../../common/helpers/error';
+import {escapeProps, generateProps} from '../helpers/props';
+import ExternalServices from '../../client/components/pages/externalService';
+import Layout from '../../client/containers/layout';
+import React from 'react';
+import ReactDOMServer from 'react-dom/server';
+import express from 'express';
+import target from '../templates/target';
+import url from 'url';
+
+
+const marginTime = 5 * 60 * 1000;
+const router = express.Router();
+
+
+router.get('/', async (req, res) => {
+	if (!req.user) {
+		return res.redirect('/register');
+	}
+	const {alertType, alertDetails} = req.query;
+	const {orm} = req.app.locals;
+	const editorId = req.user.id;
+
+	const cbUser = await orm.func.externalServiceOauth.getOauthToken(
+		editorId,
+		'critiquebrainz',
+		orm
+	);
+
+	let cbPermission = 'disable';
+	if (cbUser) {
+		cbPermission = 'review';
+	}
+
+	const props = generateProps(req, res, {
+		alertDetails,
+		alertType,
+		cbPermission
+	});
+
+	const markup = ReactDOMServer.renderToString(
+		<Layout {...propHelpers.extractLayoutProps(props)}>
+			<ExternalServices
+				{...escapeProps(props)}
+				alertDetails={props.alertDetails}
+				alertType={props.alertType}
+				cbPermission={props.cbPermission}
+			/>
+		</Layout>
+	);
+
+	return res.send(target({
+		markup,
+		props: escapeProps(props),
+		script: '/js/externalService.js',
+		title: 'External Services'
+	}));
+});
+
+
+router.get('/critiquebrainz/callback', auth.isAuthenticated, async (req, res, next) => {
+	const {orm} = req.app.locals;
+	const {code} = req.query;
+	const editorId = req.user.id;
+	if (!code) {
+		return next(new BadRequestError('Response type must be code'));
+	}
+	const token = await cbHelper.fetchAccessToken(code, editorId, orm);
+
+	let alertType = 'success';
+	let alertDetails = 'You have successfully linked your account with CritiqueBrainz.';
+	if (!token) {
+		alertType = 'danger';
+		alertDetails = 'Failed to connect to CritiqueBrainz';
+	}
```

Comment:
We already have an alert system in place, you should be able to add an alert to `res.locals.alert` like it is done here for example and remove the code to display these alerts from the ExternalServices component:
https://github.com/metabrainz/bookbrainz-site/blob/e001f4b05729528fb7b414f458bccd9457b457b6/src/server/app.js#L148

Reply:
After further discussion, the existing alert system is not entirely appropriate, as it only supports alerts sent as props on page load.
In this feature we also need to use alerts to show dynamic changes, which will require refactoring the alert system as a higher-order-component and/or using a React Context so that we can create alerts from any place in the front-end codebase.